File: android/src/main/java/io/parity/signer/components/networkicon/jdenticon/jdenticon_kotlin/Graphics.kt
package io.parity.signer.components.networkicon.jdenticon.jdenticon_kotlin

import kotlin.math.floor

internal class Graphics(renderer: Renderer) {

    val _renderer = renderer
    var _transform = Transform.noTransform()

    fun addPolygon(points: List<Point>, invert: Boolean = false) {
        var transform = this._transform
        var transformedPoints = ArrayList<Point>()
        var i = if (invert) points.size - 1 else 0
        if (!invert) {
            for (x in i until points.size) {
                transformedPoints.add(
                        transform.transformPoint(
                                points[x].x,
                                points[x].y
                        )
                );
            }
        } else {
            for (x in i downTo 0) {
                transformedPoints.add(
                        transform.transformPoint(
                                points[x].x,
                                points[x].y
                        )
                );
            }
        }
        this._renderer.addPolygon(transformedPoints)
    }

    fun addCircle(x: Float, y: Float, size: Float, invert: Boolean = false) {
        var p = this._transform.transformPoint(x, y, size, size)
        this._renderer.addCircle(p, size, invert)
    }

    fun addRectangle(x: Float, y: Float, w: Float, h: Float, invert: Boolean = false) {
        this.addPolygon(
                listOf(
                        Point(x, y),
                        Point(x + w, y),
                        Point(x + w, y + h),
                        Point(x, y + h)
                ),
                invert
        )
    }

    fun addTriangle(x: Float, y: Float, w: Float, h: Float, r: Float, invert: Boolean = false) {
        var points = ArrayList<Point>(listOf(
            Point(x + w, y),
            Point(x + w, y + h),
            Point(x, y + h),
            Point(x, y)
        ))
        //original splice method did not use point objects, and needed to delete two elements
        points.removeAt(floor(r).toInt() % 4);
        this.addPolygon(points, invert);
    }

    fun addRhombus(x: Float, y: Float, w: Float, h: Float, invert: Boolean = false) {
        this.addPolygon(
                listOf(
                        Point(x + w / 2f, y),
                        Point(x + w, y + h / 2f),
                        Point(x + w / 2f, y + h),
                        Point(x, y + h / 2f)
                ),
                invert
        )
    }
}

File: ios/Packages/Jdenticon/Sources/Jdenticon/Drawing/Graphics.swift
//
//  Graphics.swift
//
//
//  Created by Krzysztof Rodak on 23/08/2023.
//

import CoreGraphics
import Foundation

final class Graphics {
    private let renderer: Renderer
    var transform: Transform = .noTransform()

    init(renderer: Renderer) {
        self.renderer = renderer
    }

    func addPolygon(_ points: [CGPoint], _ inverted: Bool = false) {
        let currentTransform = transform
        var transformedPoints = [CGPoint]()

        let indices: [Int] =
            if inverted {
                points.indices.reversed()
            } else {
                Array(points.indices)
            }

        indices.forEach { index in
            let transformedPoint = currentTransform.transformPoint(x: points[index].x, y: points[index].y)
            transformedPoints.append(transformedPoint)
        }

        renderer.addPolygon(points: transformedPoints)
    }

    func addCircle(x: CGFloat, y: CGFloat, size: CGFloat, invert: Bool = false) {
        let transformedPoint = transform.transformPoint(x: x, y: y, w: size, h: size)
        renderer.addCircle(point: transformedPoint, diameter: size, counterClockwise: invert)
    }

    func addRectangle(x: CGFloat, y: CGFloat, w: CGFloat, h: CGFloat, invert: Bool = false) {
        let rectanglePoints = [
            CGPoint(x: x, y: y),
            CGPoint(x: x + w, y: y),
            CGPoint(x: x + w, y: y + h),
            CGPoint(x: x, y: y + h)
        ]
        addPolygon(rectanglePoints, invert)
    }

    func addTriangle(x: CGFloat, y: CGFloat, w: CGFloat, h: CGFloat, r: CGFloat, invert: Bool = false) {
        var points = [
            CGPoint(x: x + w, y: y),
            CGPoint(x: x + w, y: y + h),
            CGPoint(x: x, y: y + h),
            CGPoint(x: x, y: y)
        ]
        points.remove(at: Int(floor(r)) % 4)
        addPolygon(points, invert)
    }

    func addRhombus(x: CGFloat, y: CGFloat, w: CGFloat, h: CGFloat, invert: Bool = false) {
        let rhombusPoints = [
            CGPoint(x: x + w / 2, y: y),
            CGPoint(x: x + w, y: y + h / 2),
            CGPoint(x: x + w / 2, y: y + h),
            CGPoint(x: x, y: y + h / 2)
        ]
        addPolygon(rhombusPoints, invert)
    }
}
